beacon-chain/blockchain/forkchoice_update_execution.go

```diff
@@ -40,28 +42,58 @@ func (s *Service) getStateAndBlock(ctx context.Context, r [32]byte) (state.Beaco
 	return headState, newHeadBlock, nil
 }
 
-func (s *Service) forkchoiceUpdateWithExecution(ctx context.Context, newHeadRoot [32]byte) error {
+func (s *Service) forkchoiceUpdateWithExecution(ctx context.Context, newHeadRoot [32]byte, proposingSlot primitives.Slot) error {
 	isNewHead := s.isNewHead(newHeadRoot)
-	if !isNewHead && !s.isNewProposer() {
+	isNewProposer := s.isNewProposer(proposingSlot)
+	if !isNewHead && !isNewProposer {
 		return nil
 	}
 
-	headState, headBlock, err := s.getStateAndBlock(ctx, newHeadRoot)
-	if err != nil {
-		log.WithError(err).Error("Could not get forkchoice update argument")
-		return nil
+	var headState state.BeaconState
+	var headBlock interfaces.ReadOnlySignedBeaconBlock
+	var headRoot [32]byte
+	var err error
+
+	shouldUpdate := isNewHead
+	if isNewHead && isNewProposer && !features.Get().DisableReorgLateBlocks {
+		if proposingSlot == s.CurrentSlot() {
+			proposerHead := s.ForkChoicer().GetProposerHead()
+			if proposerHead != newHeadRoot {
+				shouldUpdate = false
+			}
+		} else if s.ForkChoicer().ShouldOverrideFCU() {
+			shouldUpdate = false
+		}
+	}
+	if shouldUpdate {
+		headRoot = newHeadRoot
+		headState, headBlock, err = s.getStateAndBlock(ctx, newHeadRoot)
+		if err != nil {
+			log.WithError(err).Error("Could not get forkchoice update argument")
+			return nil
+		}
+	} else {
+		// We are guaranteed that the head block is the parent
+		// of the incoming block. We do not process the slot
+		// because it will be processed anyway in notifyForkchoiceUpdate
+		headState = s.headState(ctx)
```

Comment:
in the event the reorg feature is disabled, should we be doing it this way ? This isn't gated by a flag and would be applied for users who want to disable it

Reply:
yes we already do that this way: that is we send an FCU with the same head as the previous slot for example if the block is missing. This is needed otherwise we would have a payloadID cache miss. However this is indeed redundant and we may be sending 2 or 3 times FCU with the same attributes here. But I do not know how to prevent this without more elaborate and bug-prone constructions. Since a repeated FCU is ignored by the EL, I think this is no problem